rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/model/PathConfigDiagnostics.java

```diff
@@ -0,0 +1,116 @@
+    }
+
+    public void publishDiagnostics(ISourceLocation project, IList messages) {
+        // Gather messages per file
+        Map<ISourceLocation, List<Diagnostic>> diagnosticsPerFile = Diagnostics.translateMessages(messages, cm);
+
+        // Publish diagnostics, but only for files for which diagnostics have changed
+        for (var entry : diagnosticsPerFile.entrySet()) {
+            ISourceLocation file = entry.getKey().top();
+            List<Diagnostic> newDiagnostics = entry.getValue();
+
+            Map<ISourceLocation, List<Diagnostic>> fileDiagsPerProject = publishedProjectDiagsPerFile.computeIfAbsent(file, (f) -> new LinkedHashMap<>());
+            @Nullable List<Diagnostic> publishedDiags = fileDiagsPerProject.get(file);
+
+            // Only publish diagnostics for a file if the diagnostics for that file have changed in the current project
+            if (!newDiagnostics.equals(publishedDiags)) {
+                // The diagnostics for this file related to the current project have changed.
+                // We need to re-publish all diagnostics for this file (including the ones from other projects)
+                fileDiagsPerProject.put(project, newDiagnostics);
+                publishFileDiagnostics(file);
+            }
+        }
+
+        // Now we have published the diagnostics of all files that occur in the list of new messages
+        // Time to remove diagnostics from files that our project previously published diagnostics for but no longer have diagnostics
+        Set<ISourceLocation> newFiles = diagnosticsPerFile.keySet().stream().map(file -> file.top()).collect(Collectors.toSet());
+        Set<ISourceLocation> filesWithoutDiags = new HashSet<>(filesWithDiagsPerProject.getOrDefault(project, Collections.emptySet()));
+        filesWithoutDiags.removeAll(newFiles);
+
+        for (ISourceLocation file : filesWithoutDiags) {
+            publishedProjectDiagsPerFile.get(file).remove(project); // File no longer has diagnostics associated with it in our project
+            publishFileDiagnostics(file);
+        }
+        filesWithDiagsPerProject.put(project, newFiles);
+    }
+
+    public void clearDiagnostics(ISourceLocation project) {
+        Set<ISourceLocation> affectedFiles = filesWithDiagsPerProject.remove(project);
+        if (affectedFiles != null) {
+            for (ISourceLocation file : affectedFiles) {
+                publishedProjectDiagsPerFile.get(file).remove(project);
+                publishFileDiagnostics(file);
+            }
+        }
+    }
+
+    private void publishFileDiagnostics(ISourceLocation file) {
+        List<Diagnostic> fileDiagnostics = new ArrayList<>();
+        for (List<Diagnostic> diags :  publishedProjectDiagsPerFile.getOrDefault(file, Collections.emptyMap()).values()) {
+            fileDiagnostics.addAll(diags);
+        }
+        client.publishDiagnostics(new PublishDiagnosticsParams(file.getURI().toString(), fileDiagnostics));
+    }
```

Comment:
the interaction between this function and the work in `publishDiagnostics` via the common fields is what makes it harder for me to review what is happening.